traffic_ops/traffic_ops_golang/user/user.go

```diff
@@ -0,0 +1,445 @@
+		"role":     validation.Validate(user.Role, validation.Required),
+		"username": validation.Validate(user.Username, validation.Required),
+		"tenantID": validation.Validate(user.TenantID, validation.Required),
+	}
+
+	// Passwords are not required for update
+	if user.LocalPassword != nil && user.ConfirmLocalPassword != nil {
+
+		if *user.LocalPassword != *user.ConfirmLocalPassword {
+			return errors.New("Passwords should match.")
+		}
+
+		_, err := auth.IsGoodLoginPair(*user.LocalPassword, *user.Username)
+		if err != nil {
+			return err
+		}
+	}
+
+	return util.JoinErrs(tovalidate.ToErrors(validateErrs))
+}
+
+func (user *TOUser) postValidate() error {
+	validateErrs := validation.Errors{
+		"localPasswd":        validation.Validate(user.LocalPassword, validation.Required),
+		"confirmLocalPasswd": validation.Validate(user.ConfirmLocalPassword, validation.Required),
+	}
+	return util.JoinErrs(tovalidate.ToErrors(validateErrs))
+}
+
+// Note: Not using GenericCreate because Scan also needs to scan tenant and rolename
+func (user *TOUser) Create() (error, error, int) {
+
+	// PUT and POST validation differs slightly
+	err := user.postValidate()
+	if err != nil {
+		return err, nil, http.StatusBadRequest
+	}
+
+	// Convert password to SCRYPT
+	*user.LocalPassword, err = auth.DerivePassword(*user.LocalPassword)
+	if err != nil {
+		return err, nil, http.StatusBadRequest
+	}
+	*user.ConfirmLocalPassword, err = auth.DerivePassword(*user.ConfirmLocalPassword)
+	if err != nil {
+		return err, nil, http.StatusBadRequest
+	}
+
+	resultRows, err := user.ReqInfo.Tx.NamedQuery(user.InsertQuery(), user)
+	if err != nil {
+		return api.ParseDBError(err)
+	}
+	defer resultRows.Close()
+
+	var id int
+	var lastUpdated tc.TimeNoMod
+	var tenant string
+	var rolename string
+
+	rowsAffected := 0
+	for resultRows.Next() {
+		rowsAffected++
+		if err = resultRows.Scan(&id, &lastUpdated, &tenant, &rolename); err != nil {
+			return nil, fmt.Errorf("could not scan after insert: %s\n)", err), http.StatusInternalServerError
+		}
+	}
+
+	if rowsAffected == 0 {
+		return nil, fmt.Errorf("no user was inserted, nothing was returned"), http.StatusInternalServerError
+	} else if rowsAffected > 1 {
+		return nil, fmt.Errorf("too many rows affected from user insert"), http.StatusInternalServerError
+	}
+
+	user.ID = &id
+	user.LastUpdated = &lastUpdated
+	user.Tenant = &tenant
+	user.RoleName = &rolename
+	user.LocalPassword = nil
+	user.ConfirmLocalPassword = nil
+
+	return nil, nil, http.StatusOK
+}
+
+// returning true indicates the data related to the given tenantID should be visible
+// this is just a linear search;`tenantIDs` is presumed to be unsorted
+func checkTenancy(tenantID *int, tenantIDs []int) bool {
+	for _, id := range tenantIDs {
+		if id == *tenantID {
+			return true
+		}
+	}
+	return false
+}
+
+// This is not using GenericRead because of this tenancy check. Maybe we can add tenancy functionality to the generic case?
```

Comment:
if `tc.User` implements the `Tenantable` interface,  this will be handled automatically by the generic handler:
https://github.com/apache/trafficcontrol/blob/master/traffic_ops/traffic_ops_golang/api/shared_handlers.go#L215

Reply:
You referenced the update handler. I believe that also applies to create and delete as well, but not read. Read won't tell you that you don't have access, rather filters out what you do have.

Reply:
+1